octopusdeploy_framework/schemas/lifecycle.go

```diff
@@ -149,3 +157,57 @@ func getRetentionPolicyAttribute() datasourceSchema.ListNestedAttribute {
 		},
 	}
 }
+
+type retentionPolicyValidator struct{}
+
+func (v retentionPolicyValidator) Description(ctx context.Context) string {
+	return "validates that should_keep_forever is true only if quantity_to_keep is 0"
+}
+
+func (v retentionPolicyValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+func (v retentionPolicyValidator) ValidateObject(ctx context.Context, req validator.ObjectRequest, resp *validator.ObjectResponse) {
+	var retentionPolicy struct {
+		QuantityToKeep    types.Int64  `tfsdk:"quantity_to_keep"`
+		ShouldKeepForever types.Bool   `tfsdk:"should_keep_forever"`
+		Unit              types.String `tfsdk:"unit"`
+	}
+
+	diags := tfsdk.ValueAs(ctx, req.ConfigValue, &retentionPolicy)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !retentionPolicy.QuantityToKeep.IsNull() && !retentionPolicy.ShouldKeepForever.IsNull() {
+		quantityToKeep := retentionPolicy.QuantityToKeep.ValueInt64()
+		shouldKeepForever := retentionPolicy.ShouldKeepForever.ValueBool()
+
+		if quantityToKeep == 0 && !shouldKeepForever {
+			resp.Diagnostics.AddAttributeError(
+				req.Path.AtName("should_keep_forever"),
+				"Invalid retention policy configuration",
+				"should_keep_forever must be true when quantity_to_keep is 0",
+			)
+		} else if quantityToKeep != 0 && shouldKeepForever {
+			resp.Diagnostics.AddAttributeError(
+				req.Path.AtName("should_keep_forever"),
+				"Invalid retention policy configuration",
+				"should_keep_forever must be false when quantity_to_keep is not 0",
+			)
+		}
+	}
+
```

Comment:
I think this needs an error check here

```suggestion
	if resp.Diagnostics.HasError() {
		return
	}
```